posts/2021-06-11-kubernetes-operator-with-java-and-quarkus.adoc

```diff
@@ -0,0 +1,722 @@
+
+   StatefulSet statefulset = kubernetesClient.apps().statefulSets().withName(spec.getName()).get();
+
+   if(statefulset == null) {
+       log.info("Create statefulset " + spec.getName());
+
+       statefulset = new StatefulSetBuilder()
+               .withNewMetadata()
+               .withName(spec.getName())
+               .endMetadata()
+               .withNewSpec()
+               .withReplicas(1)
+               .withNewTemplate()
+               .withNewMetadata()
+               .addToLabels("app", spec.getName())
+               .endMetadata()
+               .withNewSpec()
+               .addNewContainer()
+               .withName(spec.getName())
+               .withImage(spec.getImage())
+               .withCommand("sh", "-c", "while sleep 5; do cat /tmp/data.txt; done")
+               .addNewPort()
+               .withContainerPort(80)
+               .endPort()
+               .endContainer()
+               .endSpec()
+               .endTemplate()
+               .withNewSelector()
+               .addToMatchLabels("app", spec.getName())
+               .endSelector()
+               .endSpec()
+               .build();
+
+       kubernetesClient.apps().statefulSets().inNamespace(namespace).create(statefulset);
+
+       return UpdateControl.noUpdate();
+
+   } else if (statefulset.getStatus().getReplicas() < spec.getReplicas()) {
+
+       log.info("Scale statefulset up: " + spec.getName());
+
+       kubernetesClient.apps()
+               .statefulSets()
+               .inNamespace(namespace).withName(spec.getName()).scale(statefulset.getStatus().getReplicas() + 1, true);
+
+       return UpdateControl.noUpdate();
+
+   } else if (statefulset.getStatus().getReplicas() > spec.getReplicas()) {
+
+       log.info("Scale statefulset down: " + spec.getName());
+
+       kubernetesClient.apps()
+               .statefulSets()
+               .inNamespace(namespace).withName(spec.getName()).scale(spec.getReplicas(), true);
+
+       return UpdateControl.noUpdate();
+   }
+
+   return UpdateControl.noUpdate();
+----
+====
+
+Containers are running the following command to check if the data is present in the container, and outputs an error to a log if data is not present.
+
+----
+"while sleep 5; do cat /tmp/data.txt; done"
+----
+
+PodWatcher watches pod creation and modification. This functionality provides the ability to dynamically react to changes of the cluster resources.
+
+[,java]
+----
+public class PodWatcher {
+
+private final KubernetesClient kubernetesClient;
+
+void onStartup(@Observes StartupEvent startupEvent) throws IOException {
+
+   public PodWatcher(KubernetesClient kubernetesClient) {
+      this.kubernetesClient = kubernetesClient;
+   }
+
+
+   kubernetesClient.pods().watch(new Watcher<Pod>() {
+
+   @Override
+   public void eventReceived(Action action, Pod pod) {...}
+
+   @Override
+   public void onClose(WatcherException e) {...}
+});
+
+}
+----
+
+Start an operator
```

Comment:
You should probably make this a section header (H4 or H5?)

Reply:
updated some of the phrases to make then section headers